src/API/Site/Controllers/MerchantCenter/PriceBenchmarksController.php

```diff
@@ -0,0 +1,231 @@
+
+	use ContainerAwareTrait;
+
+	/**
+	 * Product helper class.
+	 *
+	 * @var ProductHelper
+	 */
+	protected $product_helper;
+
+	/**
+	 * Merchant Report constructor.
+	 *
+	 * @param RESTServer    $server
+	 * @param ProductHelper $product_helper
+	 */
+	public function __construct( RESTServer $server, ProductHelper $product_helper ) {
+		parent::__construct( $server );
+		$this->product_helper = $product_helper;
+	}
+
+	/**
+	 * Register rest routes with WordPress.
+	 */
+	public function register_routes(): void {
+		$this->register_route(
+			'mc/price-benchmarks',
+			[
+				[
+					'methods'  => TransportMethods::READABLE,
+					'callback' => $this->get_price_benchmarks_callback(),
+				],
+				'schema' => $this->get_api_response_schema_callback(),
+			]
+		);
+	}
+
+	/**
+	 * Maps query arguments from the REST request.
+	 *
+	 * @param Request $request REST Request.
+	 * @return array
+	 */
+	protected function prepare_query_arguments( Request $request ): array {
+		$args = wp_parse_args(
+			array_intersect_key(
+				$request->get_query_params(),
+				$this->get_collection_params()
+			),
+			$request->get_default_params()
+		);
+
+		return $args;
+	}
+
+	/**
+	 * Get the callback function for the price benchmarks request.
+	 *
+	 * @return callable
+	 */
+	protected function get_price_benchmarks_callback(): callable {
+		return function ( Request $request ) {
+			try {
+				/** @var MerchantPriceBenchmarks $merchant */
+				$merchant = $this->container->get( MerchantPriceBenchmarks::class );
+
+				$benchmark_data      = $merchant->get_benchmark_data( $this->prepare_query_arguments( $request ) );
+				$price_insights_data = $merchant->get_price_insights( $this->prepare_query_arguments( $request ) );
+
+				// Map the data to the required format.
+				$response_data = $this->map_price_benchmarks_response( $benchmark_data, $price_insights_data );
+
+				return new Response( $response_data );
+			} catch ( Exception $e ) {
+				return $this->response_from_exception( $e );
+			}
+		};
+	}
+
+	/**
+	 * Maps the benchmark and price insights data to the required API response format.
+	 *
+	 * @param array $benchmark_data Raw benchmark data.
+	 * @param array $price_insights_data Raw price insights data.
+	 * @return array Mapped response data.
+	 */
+	protected function map_price_benchmarks_response( array $benchmark_data, array $price_insights_data ): array {
+		$mapped_data = [];
+
+		if ( empty( $benchmark_data['results'] ) || empty( $price_insights_data['results'] ) ) {
+			return $mapped_data;
+		}
+		// Loop through the benchmark data results.
+		foreach ( $benchmark_data['results'] as $benchmark_result ) {
+			$product_view          = $benchmark_result['productView'];
+			$price_competitiveness = $benchmark_result['priceCompetitiveness'];
+
+			// Find the matching price insights data by offer_id.
+			foreach ( $price_insights_data['results'] as $price_insights_result ) {
```

Comment:
Rather than the nested loops here, let's loop through both data types independently and add them the the `$mapped_data` array keyed to the product ID, that way we end up with a single array that should contain a list like:

```php
$mapped_data = [
  '123456' => [
    'product_view' => [] // array from benchmarks
    'price_competitiveness' => [] // array from benchmarks
    'price_insights' => [] // array from price insights data
  ],
]
```

Otherwise, we're looping through the same `$price_insights_data` list for each item in the benchmark data list.

Reply:
I really missed this performance improvement, I want to make sure it gets across the finish line. I was stretched a bit last night and forgot to fix this one, but I’ve updated it now.